### PR TITLE
Fix several problems with performance annotations

### LINKED ISSUE #64511: @_noLocks is overly restrictive prohibiting common real-time code patterns
# @_noLocks is overly restrictive prohibiting common real-time code patterns

Introduction
============
I am currently working on a project using real-time audio processing code with swift. Part of this project is to evaluate using swift for real-time critical code.
To ensure real-time safe code, I am using the experimental perfomance annotations (`-experimental-performance-annotations`), in particular `@_noLocks`, and I am getting performance annotation compiler errors when writing anything slightly more complex than trivial DSP code.

Description
===========
Below is a list of the errors I am sseing when trying to do things that seem trivial to me and where I believe that the compiler should be able to understand that there are no locks necessary to generate the code. I have ordered the list in order of decreasing importance to our project (i.e. the first item is the most important etc.).
Disclaimer: Although I have expert knowledge in real-time C++ code, I am fairly new to Swift (6 months experience) and so, some things on the list below may also be due to my lack of knowledge on exactly how swift works.

List of performance annotation issues
=====================================

## 1. Referencing a member or method of another class is not possible in a @_noLocks context

Compiling the following code:

```swift
class Foo {
    public let value: Int
    
    init(value: Int) {
        self.value = value
    }
}

class Goo {
    let foo = Foo.init(value: 3)        // <-- try replacing "let" with "var"
    
    @_noLocks
    public func test(x: Int) -> Int {
        return foo.value * x            // <----- !!error here!!
    }
}
```

produces the following error:
```
Ending the lifetime of a value of type 'Foo' can cause a deallocation
``` 

I'm not sure why I'm seeing this error. Maybe this is my limited swift knowledge. I understand that, theoretically, a method in `Foo` could call back into `Goo` and de-allocate the `foo` instance. However, here the compiler has visibility into `Foo` and should understand that this is not happening.
Interestingly, replacing the `foo` memeber `let` declaration with a `var` declaration produces the following error:
```
Called function is not known at compile time and can have unpredictable performance
```
I do not see the difference with `let` or `var` here as I'm not trying to mutate the reference itself. Also, the called function is known at compile time. Note that it doesn't make a difference if `value` is a member or method. Also, strangely, this error does not seem to occur when accessing methods/members of the standard swift library or other modules.
The only current workaround I have for my code is to pack everything into a single large class which is becoming more and more unpratical as the code size grows.

## 2. Dealing with C-like arrays is clunky, unsafe and bug-prone in a @_noLocks context

Any developer using swift's performance annotations is likely going to be maniupulating elements in an array/buffer etc. This is surprisingly difficult when using `@_noLocks`. A naïve approach is to use swift Arrays:

```swift
class AccumulateWithLookupTableA {
    let lookupTable: [Float] = [0.5, 0.4, 0.3, 1.0, 0.2, 0.3]
    
    @_noLocks
    public func process(_ x: Float) -> Float {
        var sum: Float = 0.0
        
        for flt in lookupTable {    // <----- !!error here!!
            sum += flt * x
        }
        
        return sum
    }
}
```
Here, the swift compiler (somewhat understandably) does not like generating the sequence in the `for` loop and throws:

```
This code performs reference counting operations which can cause locking
```

To fix this, we need to convert the for loop into an index based loop:

```swift
class AccumulateWithLookupTableB {
    let lookupTable: [Float] = [0.5, 0.4, 0.3, 1.0, 0.2, 0.3]
    let count: Int
    
    init() {
        count = lookupTable.count
    }
    
    @_noLocks
    public func process(_ x: Float) -> Float {
        var sum: Float = 0.0
        
        for i in 0..<count {    // <---- we can't use lookupTable.count here as the compiler will throw an error
            sum += lookupTable[i] * x   // <----- !!error here!!
        }
        
        return sum
    }
}
```

Here we get the following error

```
This code performs reference counting operations which can cause locking
```
when trying to call the subscript operator of `Array`. We can try to convert the `Array` into an `UnsafeMutableBufferPointer`:

```swift
class AccumulateWithLookupTableC {
    let lookupTable: [Float] = [0.5, 0.4, 0.3, 1.0, 0.2, 0.3]
    let count: Int
    
    init() {
        count = lookupTable.count
    }
    
    @_noLocks
    public func process(_ x: Float) -> Float {
        return lookupTable.withUnsafeBufferPointer { lut in // <----- !!error here!!
            var sum: Float = 0.0
            for i in 0..<count {
                sum += lut[i] * x
            }
            
            return sum
        }
    }
}
```

The above code would be my prefered solution, however, the compiler does not like me calling `withUnsafeBufferPointer` in a `@_noLocks` context:
```
This code performs reference counting operations which can cause locking
```
Note that with the above code snippets being invalid, there is also no way for a real-time method (i.e. a method marked `@_noLocks`), to have a parameter of type `Array` as there would be no way to process this array.

We might want to try to move the `withUnsafeBufferPointer` into the constructor:

```swift
class AccumulateWithLookupTableD {
    let lookupTable: [Float] = [0.5, 0.4, 0.3, 1.0, 0.2, 0.3]
    let count: Int
    let lut: UnsafeBufferPointer<Float>
    
    init() {
        count = lookupTable.count
        lut = lookupTable.withUnsafeBufferPointer { return $0 }
    }
    
    @_noLocks
    public func process(_ x: Float) -> Float {
        var sum: Float = 0.0
        for i in 0..<count {
            sum += lut[i] * x
        }
        
        return sum
    }
}
```

This works without any warning/errors, but invites the developer to write unsafe code as `withUnsafeBufferPointer` also works if the receiver is a temporary or local variable.

My current workaround is to use `UnsafeMutableBufferPointer.allocate` instead of using an `Array`:

```swift
class AccumulateWithLookupTableE {
    let count: Int
    let lookupTable: UnsafeMutableBufferPointer<Float>
    
    init() {
        count = 6
        lookupTable = .allocate(capacity: count)
        
        lookupTable[0] = 0.5
        lookupTable[1] = 0.4
        lookupTable[2] = 0.3
        lookupTable[3] = 1.0
        lookupTable[4] = 0.2
        lookupTable[5] = 0.3
    }
    
    deinit {
        lookupTable.deallocate()    // <----- don't forget this here
    }
    
    @_noLocks
    public func process(_ x: Float) -> Float {
        var sum: Float = 0.0
        for i in 0..<count {
            sum += lookupTable[i] * x
        }
        
        return sum
    }
}
```

However, here the developer must not forget to call `deallocate` to ensure no memory leaking. Also, this restricts my `@_noLocks` methods to only take `UnsafeBufferPointer` typed parameters instead of having `Array` typed parameters, as there is no way to access the latter in a `@_noLocks` context.

## 3. No warning/error-free way to access a variable length array in a C struct

In the following code, I'm trying to iterate through a variable length array (`mBuffers`) which resides in a C struct:

```c++
struct AudioBufferList
{
    UInt32      mNumberBuffers;
    AudioBuffer mBuffers[1]; // this is a variable length array of mNumberBuffers elements
};
```

This struct is a fundamental part of the CoreAudioKit API (see [here](https://developer.apple.com/documentation/coreaudiotypes/audiobufferlist?language=objc)). swift translates this C-struct into

```swift
public struct AudioBufferList {
    public var mNumberBuffers: UInt32
    public var mBuffers: (AudioBuffer) // this is a variable length array of mNumberBuffers elements
}
```

The most elegant way to do this is with this code:

```swift
class Dummy {
    private let bufferKeyPath = \AudioBufferList.mBuffers
    
    @_noLocks
    func getNumberOfChannelsOfLastBufferSafe(_ bufferList: UnsafeMutablePointer<AudioBufferList>) -> UInt32 {
        let numBuffers = Int(bufferList.pointee.mNumberBuffers)
        let buffers = UnsafeBufferPointer<AudioBuffer>(start: bufferList.pointer(to: bufferKeyPath), count: numBuffers) // <--- !!error here!!
        return buffers[numBuffers - 1].mNumberChannels
    }
```

However, the above code gives us the following error:

```
Called function is not known at compile time and can have unpredictable performance
```

The only workaround I found is the following unsafe code (which is also an accepted [stack-overflow answer](https://stackoverflow.com/questions/27061028/swift-access-to-variable-length-array)):

```swift
@_noLocks
func getNumberOfChannelsOfLastBufferUnsafe(_ bufferList: UnsafeMutablePointer<AudioBufferList>) -> UInt32 {
    let numBuffers = Int(bufferList.pointee.mNumberBuffers)
    let buffers = UnsafeBufferPointer<AudioBuffer>(start: &bufferList.pointee.mBuffers, count: numBuffers) // <---- !!warning here!!
    return buffers[numBuffers - 1].mNumberChannels
}
```
Here, the compiler rightly generates the warning:

```
Initialization of 'UnsafeBufferPointer<AudioBuffer>' results in a dangling buffer pointer
```

as swift is allowed to de-allocate the pointer created by the `&` operator at anytime (unlike C/C++).

## 4. rounding numeric values is not possible in a @_noLocks context

The following code:

```swift
@_noLocks
func brokenRounding(_ flt: Float) -> Int {
    return Int(round(flt))
}
```

causes the following compiler error:

```
<unknown>:0: error: Using type 'FloatingPointRoundingRule' can cause metadata allocation or locks
```

Notice in the above error, how the compiler is even having trouble figuring out which file/line caused the error.

## 5. bit-shifting is not possible in a @_noLocks context

The followig code:

```swift
@_noLocks
func testBitShift(_ x: Int) -> Int {
    return x << 1
}
```

causes the following compiler error:

```
Using type 'Int' can cause metadata allocation or locks
```

## 6. Using enums/flags is overly complicated in a @_noLocks context

Consider the following `OptionSet` which was generated from a C/C++ bit flag set:

```swift
struct ShippingOptions: OptionSet {
    let rawValue: Int

    public static var nextDay:   ShippingOptions { get { return ShippingOptions(rawValue: 1) }}
    public static var secondDay: ShippingOptions { get { return ShippingOptions(rawValue: 2) }}
    public static var priority:  ShippingOptions { get { return ShippingOptions(rawValue: 4) }}
    public static var standard:  ShippingOptions { get { return ShippingOptions(rawValue: 8) }}
}
```
Note: the original code was using bit-shifting for the values but this results in a compiler error (see list point no. 5 above).

The following code:

```swift
@_noLocks
func shippingOptionsIsSecondDay(_ options: ShippingOptions) -> Bool {
    return options.contains(.secondDay)
}
```

causes a compiler error:

```
This code pattern can cause metadata allocation or locks
```

This can be resolved with the following workaround

```swift
@_noLocks
func hasSampleTimeWorkaround(_ flags: AudioTimeStampFlags) -> Bool {
    return (flags.rawValue & AudioTimeStampFlags.sampleTimeValid.rawValue) != 0
}
```

but is unnecessarily verbose.

Similar, counting the number of enum values:

```swift
public enum Weekdays: Int, CaseIterable {
    case monday = 0, tuesday, wednesday, thursday, friday, saturday, sunday
}

@_noLocks
func noCounting() -> Int {
    return Weekdays.allCases.count  // <---- !!error here!!
}
```

causes the following error:

```
Called function is not available in this module and can have unpredictable performance
```

The workaround is to cache the count in a non-`@_noLocks` context (for example, a constructor) and then used the cached count.

## Attachment
Attached is a swift file `noLockErrors.zip` which contains all of these examples in a single file.

**Environment**
- Swift compiler version info: Apple Swift version 5.9-dev (LLVM 43db6a5b91dd225, Swift 43f0694fa3e0924), Target: arm64-apple-macosx13.0
- Xcode version info: Xcode 14.2 Build version 14C18
- Deployment target: macOS 13.1

@eeckstein 
[noLockErrors.zip](https://github.com/apple/swift/files/11033239/noLockErrors.zip)

### PR DESCRIPTION
**Pull request by eeckstein**
And rewrite the MandatoryGenericSpecializer in Swift.
The new MandatoryPerformanceOptimizations pass is more powerful because it uses the existing instruction simplification infrastructure. This fixes some false alarms with the performance annotations `_noLocks` and `_noAllocation`.

This PR contains several other small fixes and improvements, like extracting the named-return-value optimization from CopyForwarding into its own pass.
For details see the commit messages.

rdar://107576739
rdar://107576844
rdar://107017103, https://github.com/apple/swift/issues/64511

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: lib/SILOptimizer/PassManager/PassPipeline.cpp [hunk 1]
```diff
@@ -758,6 +758,11 @@ static void addLowLevelPassPipeline(SILPassPipelinePlan &P) {
 
   addFunctionPasses(P, OptimizationLevelKind::LowLevel);
 
+  // The NamedReturnValueOptimization shouldn't be done before serialization,
+  // because it might prevent predictable memory optimizations in the caller
+  // after inlining.
```
Comment (atrick) [inline lib/SILOptimizer/PassManager/PassPipeline.cpp:764]: Thanks for commenting this pipeline restriction. Pipeline comments are helpful, but I think we should also have a convention that pass dependencies are specified in the file that defines the pass. That's where people will see it when they are investigating the pass, and the information won't be lost as easily. I regularly see the pipeline comments get messed up as people hack on it. In the pass file, you can describe the dependency in more detail with examples. In this case, I'm not sure why NRVO interferes with predictable mem opts, but deferring it until after serialization seems reasonable. It doesn't seem needed for canonicalization and might not be useful if inlining happens.

**Review by atrick — APPROVED** [review]
Looks great based on a quick review. I mostly need to trust you on correctness. Too much to review in careful detail.

File: test/SILOptimizer/mandatory_performance_optimiations.sil [hunk 2]
```diff
@@ -0,0 +1,132 @@
+// RUN: %target-sil-opt -enable-sil-verify-all %s -mandatory-performance-optimizations | %FileCheck %s
```
Comment (nate-chandler) [inline test/SILOptimizer/mandatory_performance_optimiations.sil:1]: Nit: filename typo `optimiations`

File: lib/SILOptimizer/PassManager/PassPipeline.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SILOptimizer/PassManager/PassPipeline.cpp:764]: 👍I added a comment

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Pull request merged**